dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/preProcess/update/ProgramInstanceUpdatePreProcessor.java

```diff
@@ -0,0 +1,150 @@
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import static org.hisp.dhis.event.EventStatus.ACTIVE;
+import static org.hisp.dhis.event.EventStatus.COMPLETED;
+import static org.hisp.dhis.event.EventStatus.SCHEDULE;
+import static org.hisp.dhis.event.EventStatus.SKIPPED;
+import static org.hisp.dhis.user.User.getSafeUsername;
+import static org.hisp.dhis.user.UserCredentials.USERNAME_MAX_LENGTH;
+import static org.hisp.dhis.util.DateUtils.parseDate;
+import static org.springframework.util.StringUtils.isEmpty;
+
+import java.util.Date;
+
+import org.hisp.dhis.category.CategoryOptionCombo;
+import org.hisp.dhis.dxf2.common.ImportOptions;
+import org.hisp.dhis.dxf2.events.event.Event;
+import org.hisp.dhis.dxf2.events.event.preProcess.PreProcessor;
+import org.hisp.dhis.dxf2.events.event.validation.ValidationContext;
+import org.hisp.dhis.dxf2.importsummary.ImportConflict;
+import org.hisp.dhis.dxf2.importsummary.ImportSummary;
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.util.DateUtils;
+
+public class ProgramInstanceUpdatePreProcessor
+    implements
+    PreProcessor
+{
+    @Override
+    public void process( final Event event, final ValidationContext ctx )
+    {
+        final ProgramStageInstance programStageInstance = ctx.getProgramStageInstanceMap().get( event.getEvent() );
+        final ImportOptions importOptions = ctx.getImportOptions();
+        final OrganisationUnit organisationUnit = ctx.getOrganisationUnitMap().get( event.getUid() );
+        final CategoryOptionCombo categoryOptionCombo = ctx.getCategoryOptionComboMap().get( event.getUid() );
+
+        Date dueDate = new Date();
+        if ( event.getDueDate() != null )
+        {
+            dueDate = parseDate( event.getDueDate() );
+        }
+
+        if ( event.getEventDate() != null )
+        {
+            programStageInstance.setExecutionDate( parseDate( event.getEventDate() ) );
+        }
+
+        if ( categoryOptionCombo != null )
+        {
+            programStageInstance.setAttributeOptionCombo( categoryOptionCombo );
+        }
+
+        final String storedBy = getValidUsername( event.getStoredBy(), null,
+            importOptions.getUser() != null ? importOptions.getUser().getUsername() : "[Unknown]" );
+
+        final String completedBy = getValidUsername( event.getCompletedBy(), null,
+            importOptions.getUser() != null ? importOptions.getUser().getUsername() : "[Unknown]" );
+
```

Comment:
I have created some shared super-class code to get this kind of data, we can refactor later